[PWA] Add a guide on offline and background operation (#25305)
mdn/content: files/en-us/web/progressive_web_apps/guides/offline_and_background_operation/index.md

---
title: Offline and background operation

> slug: Web/Progressive_web_apps/Guides/Offline_and_background_operation

Reviewer: page-type ?

Author: I think we will add page-type for this and the rest when we get around to that bit of the docs in the page-type project.

---

Usually, websites are very dependent on both reliable network connectivity and on the user having their pages open in a browser. Without network connectivity, most websites are just unusable, and if the user does not have the site open in a browser tab, most websites are unable to do anything.

However, consider the following scenarios:

- A music app enables the user to stream music while online, but can download tracks in the background and then continue to play while the user is offline.
- The user composes a long email, presses "Send", and then loses network connectivity. The device sends the email in the background, as soon as the network is available again.
- The user's chat app receives a message from one of their contacts, and although the app is not open, it displays a badge on the app icon to let the user know they have a new message.

> These are the kinds of features that users expect from platform-specific apps. If PWAs are to compete with platform-specific apps, they need to support scenarios like this. In this guide, we'll introduce a set of technologies which enable a PWA to:

Reviewer: I know this is an irritating correction. In this case reads better to me.

```suggestion
These are the kinds of features that users expect from platform-specific apps. If PWAs are to compete with platform-specific apps, they need to support scenarios like this. In this guide, we'll introduce a set of technologies that enable a PWA to:
```

> These are the kinds of features that users expect from platform-specific apps. If PWAs are to compete with platform-specific apps, they need to support these kinds of scenarios. In this guide, we'll introduce a set of technologies that enable a PWA to:

Reviewer: I don't feel strongly but another slight rewording along these lines ^^^'

```suggestion
These are the kinds of features that users expect from both native apps and PWAs. In this guide, we'll introduce a set of technologies that enable a PWA to:
```

> These are the kinds of features that users expect from platform-specific apps. If PWAs are to compete with platform-specific apps, they need to support these kinds of scenarios. In this guide, we'll introduce a set of technologies that enable a PWA to:

Author: I just went with "installed apps" -> https://github.com/mdn/content/pull/25305/commits/c75943c6072c5c0587a47f8cf6f272f5db16b87b.

- provide a good user experience even when the device has intermittent network connectivity
- update its state when the app is not running

> - notify the user about important events that have happened while the app was not running.

Reviewer: Don't we need a terminating comma on the early bullets for correct English? The under construction bullet guidelines would say you  can do each of these like a phrase or all like sentences, but in the end you must have either all periods or no periods at the end of each point.

Author: I have no idea really what the rules are with bullets, but the argument against using caps and periods for each one is that they are not complete sentences. I'm happy-ish to follow our standard for this as and when we have any. IME we are very inconsistent about this in MDN right now.

Reviewer: Yes we are inconsistent, which is why there is an attempt to make this consistent in https://github.com/mdn/content/pull/25191 - despite the heading this also touches bullet lists.

The proposed rule is that a phrase (incomplete sentence) is capitalized but with no period. A full sentence is capitalized and with period. Write to ensure a list is all of one type or the other. 
I wanted capitalisation and no period in all cases except for code entities, but not much traction gained on that.

Author: -> https://github.com/mdn/content/pull/25305/commits/26dca958226d70e4058e10df0649816073c66525

> The techologies introduced in this guide are:

Reviewer: ```suggestion
The technologies introduced in this guide are:
```

- [Service Worker API](/en-US/docs/Web/API/Service_Worker_API)
- [Background Synchronization API](/en-US/docs/Web/API/Background_Synchronization_API)
- [Background Fetch API](/en-US/docs/Web/API/Background_Fetch_API)
- [Periodic Background Synchronization API](/en-US/docs/Web/API/Web_Periodic_Background_Synchronization_API)
- [Push API](/en-US/docs/Web/API/Push_API)
- [Notifications API](/en-US/docs/Web/API/Notifications_API)

## Offline operation

> Offline operation means that the PWA can give as good an experience as possible when the device does not have network connectivity. The foundation of this, and of the other technologies in this guide, is the service worker.

Reviewer: ```suggestion
Offline operation allows a PWA to provide a good user experience even when the device does not have network connectivity. The foundation of this, and of the other technologies in this guide, is the service worker.
```

A [worker](/en-US/docs/Web/API/Web_Workers_API) is a part of a web app that runs in a separate thread to the main JavaScript code. The main code creates the worker, passing in a URL to the worker's script. The worker and the main code can't directly access each other's state, but can communicate by sending each other messages. Workers can be used to run computationally expensive tasks in the background: because they run in a separate thread, the main JavaScript code in the app, that implements the app's UI, can stay responsive to the user.

> A [service worker](/en-US/docs/Web/API/Service_Worker_API) is a specific type of worker that _controls_ some subset of the pages under the app origin (by default, all pages under the app origin).

Author: I think explaining what `scope` is, is too much detail for this guide. Could we just say:

> A [service worker](/en-US/docs/Web/API/Service_Worker_API) is a specific type of worker that _controls_ some or all of the app's pages.

?

> A [service worker](/en-US/docs/Web/API/Service_Worker_API) is a specific type of worker that _controls_ some subset of the pages under the app origin (by default, all pages under the app origin).

Reviewer 2: Sure, that works.

> A [service worker](/en-US/docs/Web/API/Service_Worker_API) is a specific type of worker. It is said to _control_ some subset of the pages under the app origin (by default, all pages under the app origin).

Reviewer: I think this is a better way of implying that the meaning of control will be expanded on below.

```suggestion
A [service worker](/en-US/docs/Web/API/Service_Worker_API) is a specific type of worker that _controls_ some subset of the pages under the app origin (by default, all pages under the app origin).
```

1. If the requested resource exists in the cache, get the resource from the cache and return the resource to the app.
2. If the requested resource does not exist in the cache, try to fetch the resource from the network.
   1. If the resource could be fetched, add the resource to the cache for next time, and return the resource to the app.
   2. If the resource could not be fetched, return some default fallback resource.

The following code sample shows an implementation of this:

```js
const putInCache = async (request, response) => {
  const cache = await caches.open("v1");
  await cache.put(request, response);
};

const cacheFirst = async ({ request, fallbackUrl }) => {
  // First try to get the resource from the cache.
  const responseFromCache = await caches.match(request);
  if (responseFromCache) {
    return responseFromCache;
  }

  // If the response was not found in the cache,
  // try to get the resource from the network.
  try {
    const responseFromNetwork = await fetch(request);
    // If the network request succeeded, clone the response:
    // - put one copy in the cache, for the next time

>     // - return the other copy to the app

Author: -> https://github.com/mdn/content/pull/25305/commits/1676093eb4c6e6519d52421f54efca02447d969e.

## Background operation

While offine operation is probably the most common use for service workers, they also enable a PWA to operate even while the main app is closed. This is possible because the service worker is allowed to run while the main app is not running.

This doesn't mean service workers run all the time: browsers may close service workers when they think it is appropriate. For example, if a service worker has been inactive for a while, it will be closed. However, the browser will restart the service worker when an event has happened that it needs to take care of. This enables a PWA to implement background operations in the following way:

- in the main app, register a request for the service worker to perform some operation
- at the appropriate time, the service worker will be restarted if necessary, and an event will fire in the service worker's scope
- the service worker will perform the operation

In the next sections, we'll discuss a few different features that use this pattern to enable a PWA to work while the main app isn't open.

> ## Background sync

Author: Yes, they should, and in my draft they were. The trouble is, the MDN ToC only includes H2 headings, so it's very hard to navigate to H3s. There is an issue to include H3 headings in the ToC: https://github.com/mdn/yari/issues/7198 but it isn't very active.

So right now we have a choice:
- have a logically sensible outline, with sections that user's can't find
- have a non-sensible outline, but sections that users can find

> ## Background sync

Reviewer 2: Gotcha, thanks for the context. In that case, I vote for keeping them as h2s, since they're very important in this article and people might want to jump to them easily.

Suppose a user composes an email and presses "Send". In a traditional website, they must keep the tab open until the app has sent the email: if they close the tab, or the device loses connectivity, then the message will not be sent. Background sync, defined in the [Background Synchronization API](/en-US/docs/Web/API/Background_Synchronization_API), is the solution to this problem for PWAs.

Background sync enables the app to ask its service worker to perform a task on its behalf. As soon as the device has network connectivity, the browser will restart the service worker, if necessary, and fire an event named [`sync`](/en-US/docs/Web/API/ServiceWorkerGlobalScope/sync_event) in the service worker's scope. The service worker can then attempt to execute the task. If the task can't be completed, then the browser may retry a limited number of times by firing the event again.

### Registering a sync event

> To ask the service worker to perform a task, the main app can access {{domxref("ServiceWorkerContainer/ready", "navigator.serviceWorker.ready")}}, which resolves with a {{domxref("ServiceWorkerRegistration")}} object. The app then calls {{domxref("SyncManager/register", "sync.register()")}} on the `ServiceWorkerRegistration` object, like this:

Author: In the end it felt best to have a dedicated section at the start introducing workers and the architecture, then consistently using comments to indicate which scope the code belonged in: https://github.com/mdn/content/pull/25305/commits/934dd5ce25fb61d3fea87c0ba013dcd2588e1a28.

```js
// main.js
async function registerSync() {
  const swRegistration = await navigator.serviceWorker.ready;
  swRegistration.sync.register("send-message");
}
```

Note that the app passes a name for the task: `"send-message"` in this case.

### Handling a sync event

As soon as the device has network connectivity, the `sync` event fires in the service worker scope. The service worker checks the name of the task and runs the appropriate function, `sendMessage()` in this case:

```js
// service-worker.js
self.addEventListener("sync", (event) => {
  if (event.tag == "send-message") {
    event.waitUntil(sendMessage());
  }
});
```

Note that we pass the result of the `sendMessage()` function into the event's {{domxref("ExtendableEvent/waitUntil", "waitUntil()")}} method. The `waitUntil()` method takes a {{jsxref("Promise")}} as a parameter and asks the browser not to stop the service worker until the promise has settled. This is also how the browser knows whether the operation succeeded or not: if the promise rejects, then the browser may retry by firing the `sync` event again.

> The `waitUntil()` method is not a guarantee that the browser will not stop the service worker: if the operation takes too long, the service worker will be stopped anyway.

Reviewer: Do we need to think here about what happens next? i.e. is this code killed, or would the service worker continue running from here if there was another sync event?

Reviewer: Or to put it another way, what do we mean by "stop"?

Author: Good point! I have attempted to clarify this, hopefully correctly!

-> https://github.com/mdn/content/pull/25305/commits/89c42c5c55fc11ccece5d95b850dc56c39583a31.

Reviewer: Yes, this is a good fix.

> ## Background fetch

Reviewer 2: The following web.dev article is a great hands-on article about downloading big files for offline viewing: https://web.dev/pwa-with-offline-streaming/

It would be good to add it to a See Also section, along with other useful links about background sync, periodic syncs, push, etc.

Author: Good idea.

Added a See also -> https://github.com/mdn/content/pull/25305/commits/4ee4cda92547660bf6e0fe81cb649544aaab174d .

```js
// main.js
async function requestBackgroundFetch(movieData) {
  const swRegistration = await navigator.serviceWorker.ready;
  const fetchRegistration = await swRegistration.backgroundFetch.fetch(
    "download-movie",
    ["/my-movie-part-1.webm", "/my-movie-part-2.webm"],
    {
      icons: movieIcons,
      title: "Downloading my movie",
      downloadTotal: 60 * 1024 * 1024,
    }
  );
  //...
}
```

We're passing three arguments into `backgroundFetch.fetch()`:

1. an identifier for this fetch request
2. an array of {{domxref("Request")}} objects or URLs. A single background fetch request can include multiple network requests.
3. an object containing data for the UI that the browser uses to show the existence and progress of the request.

The `backgroundFetch.fetch()` call returns a {{jsxref("Promise")}} that resolves to a {{domxref("BackgroundFetchRegistration")}} object. This enables the main app to update its own UI as the request progresses. However, if the main app is closed, the fetch will continue in the background.

> The browser will display a persistent UI element reminding the user that the request is ongoing, giving them the chance to find out more about the request and cancel it if they wish. The UI will include an icon and title taken from the `icons` and `title` arguments, and uses `downloadTotal` as an estimate of the total download size, to show the request's progress.

Reviewer 2: Does this happen for installed PWAs though? 

My recollection, from when I tested with https://kinoweb.dev/, is that the browser which was used to install the PWA doesn't display any download UI.

My point is also a bit more general here, and not only about this PR. We keep talking about the browser in our PWA docs, because that's the thing that runs the app. But in many cases, we really only care about installed PWAs that are presented in their own standalone windows, where the browser is nowhere to be seen.

I wonder if we should fine tune our language here to make it clearer when we're talking about an app running in a browser tab, or an app running in its own window. And make it clearer when we're talking about the browser "engine", which is running both these types of apps, but in the background, or when we're talking about the browser chrome.

Sorry for being super vague here, but I'm trying to put myself in the shoes of a total beginner who would likely get confused by the word browser, when really, they're building a standalone app.

Author: > Does this happen for installed PWAs though?
> 
> My recollection, from when I tested with https://kinoweb.dev/, is that the browser which was used to install the PWA doesn't display any download UI.

I confess, I hadn't tested this. I just tried with kinoweb and on desktop it seems like it doesn't do background fetch at all - if I start a download and then close the app, the download is suspended. I didn't try on Android yet (it's a bit more complicated for me to do that).

But this is very surprising to me! Everything I have read, including the web.dev article, the MDN reference, and the specification, seems to indicate that this is a requirement:

- https://developer.chrome.com/blog/background-fetch/
- https://developer.mozilla.org/en-US/docs/Web/API/Background_Fetch_API#concepts_and_usage
- https://wicg.github.io/background-fetch/#privacy-and-bandwidth-use
- https://wicg.github.io/background-fetch/#background-fetch-display

What am I missing?
 
> My point is also a bit more general here, and not only about this PR. We keep talking about the browser in our PWA docs, because that's the thing that runs the app. But in many cases, we really only care about installed PWAs that are presented in their own standalone windows, where the browser is nowhere to be seen.
> 
> I wonder if we should fine tune our language here to make it clearer when we're talking about an app running in a browser tab, or an app running in its own window. And make it clearer when we're talking about the browser "engine", which is running both these types of apps, but in the background, or when we're talking about the browser chrome.
> 
> Sorry for being super vague here, but I'm trying to put myself in the shoes of a total beginner who would likely get confused by the word browser, when really, they're building a standalone app.

Yeah, I can see this being confusing. I mean technically it is "the browser" even if the app is installed, right? Because the browser is acting as a kind of virtual machine. At least that's my understanding! But I can see it being confusing. Do you think "device" would be OK here?

This takes advantage of the fact that on a mobile device especially, connectivity is not poor so much as _intermittent_: by taking advantage of the times that the device has connectivity, the app can smooth over the connectivity gaps.

### Registering a periodic sync event

The code for registering a periodic sync event follows the same pattern as that for [registering a sync event](#registering_a_sync_event). The {{domxref("ServiceWorkerRegistration")}} has a {{domxref("ServiceWorkerRegistration.periodicSync", "periodicSync")}} property, which has a {{domxref("PeriodicSyncManager/register", "register()")}} method taking the name of the periodic sync as a parameter.

However, `periodicSync.register()` takes an extra argument, which is an object with a `minInterval` property. This represents the minimum interval, in milliseconds, between synchronization attempts:

```js
// main.js
async function registerPeriodicSync() {
  const swRegistration = await navigator.serviceWorker.ready;
  swRegistration.periodicSync.register("update-news", {
    // try to update every 2 hours

>     minInterval: 2 * 60 * 60 * 1000,

Reviewer 2: If I'm not mistaken, chromium will only let an app periodically update at most once every 24 hours. It also depends on whether the app is used on a regular basis. So this example should be changed to 24 hours I think.

Author: -> https://github.com/mdn/content/pull/25305/commits/d2bfeb5491e22b9f21277c70175308a32d4246d6

  });
}
```

### Handling a periodic sync event

Although the PWA asks for a particular interval in the `register()` call, it's up to the browser how often to generate periodic sync events. Apps that users open and interact with often will be more likely to receive periodic sync events, and will receive them more often, than apps which the user rarely or never interacts with.

When the browser has decided to generate a periodic sync event, the pattern is again familiar: it starts the service worker, if necessary, and fires a {{domxref("ServiceWorkerGlobalScope.periodicsync_event", "periodicSync")}} event in the service worker's global scope.

> The service worker's event handler checks the name of the event, and calls the appropriate function inside the event's {{domxref("ExtendableEvent/waitUntil", "waitUntil()")}} method

Reviewer 2: ```suggestion
The service worker's event handler checks the name of the event, and calls the appropriate function inside the event's {{domxref("ExtendableEvent/waitUntil", "waitUntil()")}} method:
```

```js
self.addEventListener("periodicsync", (event) => {
  if (event.tag === "update-news") {
    event.waitUntil(updateNews());
  }
});
```

> Inside `updateNews()`, the service worker can fetch and cache the latest stories. The `updateNews()` function should complete relatively quickly: if the service worker takes too long updating its content, the browser will stop it.

Reviewer: The only thing showing on a quick scan is https://chromium.googlesource.com/chromium/src/+/master/docs/security/service-worker-security-faq.md#do-service-workers-live-forever

This says that the service worker lifetime is pretty much up to the browser and browsre specific. For chrome the worker gets closed if it is unresponsive for more than 30 seconds, and if it doesn't settle its requests within 5 minutes. So I would "guess" that the time is not particular restrictive - what you should not do though is run blocking synchronous code or so many requests that the fetches are impossible even on a good network. On a rubbish network your service worker would close anyway, and that's to be expected/OK - you just don't want the service worker to do something that can never succeed even in good conditions.

> Inside `updateNews()`, the service worker can fetch and cache the latest stories. The `updateNews()` function should complete relatively quickly: if the service worker takes too long updating its content, the browser will stop it.

Author: Thank you Hamish, I added this here: https://github.com/mdn/content/pull/25305/commits/9108c46568260d83e1536b212465b7a9d17e693c .

### Unregistering a periodic sync

When the PWA no longer needs periodic background updates, (for example, because the user has switched them off in the app's settings) then the PWA should ask the browser to stop generation periodic sync events, by calling the {{domxref("PeriodicSyncManager/unregister", "unregister()")}} method of {{domxref("serviceWorkerRegistration.periodicSync", "periodicSync")}}:

```js
// main.js
async function registerPeriodicSync() {
  const swRegistration = await navigator.serviceWorker.ready;
  swRegistration.periodicSync.unregister("update-news");
}
```

## Push

The [Push API](/en-US/docs/Web/API/Push_API) enables a PWA to receive messages pushed from the server, whether the app is running or not. When the message is received by the device, the app's service worker is started and handles the message, and a [notification](/en-US/docs/Web/API/Notifications_API) is shown to the user (the specification allows for "silent push" in which no notification is shown, but no browsers support this).

> Displaying a notification to the user distracts them from whatever they were doing and has the potential to be very annoying, so use push messages with care. In general, they are suitable for situations in which you need to alert the user about something, and can't wait until the next time they open your app.

Reviewer: One real "for example" might be nice.

Reviewer 2: I think most chat apps do push notifications. In fact they very often nag you after installation to enable push notifications.

9. If the user clicks the notification or closes it, the {{domxref("ServiceWorkerGlobalScope.notificationclick_event", "notificationclick")}} and {{domxref("ServiceWorkerGlobalScope.notificationclose_event", "notificationclose")}}, respectively, are fired in the service worker's global scope. These enable the app to handle the user's response to the notification.

## Permissions and restrictions

Browsers have to find a balance in which they can provide powerful APIs to web developers while protecting users from malicious, exploitative, or poorly-written websites. One of the main protections they offer is that users can close the website's pages, and it's then not active on their device any more. The APIs described in this article tend to violate that assurance, so browsers have to take extra steps to help ensure that users are aware of this, and that the APIs are used in ways that align with the interests of users.

In this section we'll outline these steps. Several of these APIs require explicit [user permission](/en-US/docs/Web/API/Permissions_API), and various other restrictions and design choices to help protect users.

> - The Background Sync API does not need an explicit user permission, but issuing a background sync request may only be made while the main app is open, and the broweser limits the number of retries and the length of time background sync operations can take.

Reviewer: ```suggestion
- The Background Sync API does not need an explicit user permission, but issuing a background sync request may only be made while the main app is open, and browsers limit the number of retries and the length of time background sync operations can take.
```

> - The Background Fetch API requires the `"background-fetch"` permission, and the browser displays the ongoing progress of the fetch operation, enabling the user to cancel it.

Reviewer: Is this a [Permissions-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy) permission, a user-grantable permission, or both?
- Worth saying and linking to the top level page
- I don't see any documentation on this or `periodic-background-sync` - is there a plan to create it? I do see `background-sync` here: https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API
- I've been trying to create a proper target for the Permissions API versions of the permissions to link to - for example for midi, the link should be here https://developer.mozilla.org/en-US/docs/Web/API/Navigator/requestMIDIAccess#security_requirements or https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API#security_requirements - perhaps not in this PR but I think this pattern is good to follow.

Reviewer: Same question for the others below

Author: > Is this a [Permissions-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy) permission, a user-grantable permission, or both?
>
>    Worth saying and linking to the top level page

Doesn't it do this in the paragraph before the bullet list?

> In this section we'll outline these steps. Several of these APIs require explicit [user permission](/en-US/docs/Web/API/Permissions_API), and various other restrictions and design choices to help protect users.

Do we need to repeat it for each bullet?

Yes, we should make sure permissions are properly documented in the reference pages. Should we wait for https://github.com/openwebdocs/project/issues/146, when we might know where to put it? I don't think this page should have to block on that, either way.

Reviewer: Yes you do say it is a user-permission - sorry. Yes adding permissions should not block this PR.

I might still explicitly add the permission type. It does feel repetitive, but it is completely unambiguous.

```suggestion
- The Background Fetch API requires the `"background-fetch"` user permission, and the browser displays the ongoing progress of the fetch operation, enabling the user to cancel it.
```

Just my two bits, I would not wait on https://github.com/openwebdocs/project/issues/146. I'd copy the model of https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API#security_requirements and add a "Security Requirements" section to https://developer.mozilla.org/en-US/docs/Web/API/Background_Fetch_API (and others) that mentions the permissions.

Why? Well, you know the security considerations already so it is best to catch them now where they are most likely to be seen in reference. Even if I'm wrong about where they end up, I know that if you follow a pattern like this it will be easier to move to the "final" version when that is agreed. Because this is the kind of pattern people have been following, if at all. By people I mean me and @Elchi3

Author: > Yes you do say it is a user-permission - sorry. Yes adding permissions should not block this PR.
> 
> I might still explicitly add the permission type. It does feel repetitive, but it is completely unambiguous.
> 

Fair enough -> https://github.com/mdn/content/pull/25305/commits/3b54d56409d654e84c5e5369e2919cc53bd43f2f.

> Just my two bits, I would not wait on [openwebdocs/project#146](https://github.com/openwebdocs/project/issues/146). I'd copy the model of https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API#security_requirements and add a "Security Requirements" section to https://developer.mozilla.org/en-US/docs/Web/API/Background_Fetch_API (and others) that mentions the permissions.
> 
> Why? Well, you know the security considerations already so it is best to catch them now where they are most likely to be seen in reference. Even if I'm wrong about where they end up, I know that if you follow a pattern like this it will be easier to move to the "final" version when that is agreed. Because this is the kind of pattern people have been following, if at all. By people I mean me and @Elchi3

Yeah, makes sense. But not in this PR. Doing this work has made me realise how important having consistent docs for permissions and other security requirements is.

Reviewer: > Doing this work has made me realise how important having consistent docs for permissions and other security requirements is.

Yes, I had a similar epiphany when looking at Web Midi API and Permissions API. Hopefully the project codifies this soon.

The service worker's event handler checks the name of the event, and calls the appropriate function inside the event's {{domxref("ExtendableEvent/waitUntil", "waitUntil()")}} method:

```js
// service-worker.js

self.addEventListener("periodicsync", (event) => {
  if (event.tag === "update-news") {
    event.waitUntil(updateNews());
  }
});
```

Inside `updateNews()`, the service worker can fetch and cache the latest stories. The `updateNews()` function should complete relatively quickly: if the service worker takes too long updating its content, the browser will stop it.

### Unregistering a periodic sync

When the PWA no longer needs periodic background updates, (for example, because the user has switched them off in the app's settings) then the PWA should ask the browser to stop generation periodic sync events, by calling the {{domxref("PeriodicSyncManager/unregister", "unregister()")}} method of {{domxref("serviceWorkerRegistration.periodicSync", "periodicSync")}}:

```js
// main.js

async function registerPeriodicSync() {
  const swRegistration = await navigator.serviceWorker.ready;
  swRegistration.periodicSync.unregister("update-news");
}
```

## Push

> The [Push API](/en-US/docs/Web/API/Push_API) enables a PWA to receive messages pushed from the server, whether the app is running or not. When the message is received by the device, the app's service worker is started and handles the message, and a [notification](/en-US/docs/Web/API/Notifications_API) is shown to the user (the specification allows for "silent push" in which no notification is shown, but no browsers support this).

Reviewer 2: I had no idea that the push event and the user notification were tied like this. I thought a service worker could very well receive push notifications from a server without, then, displaying notifications. In my head, this was useful to display badges, or just update something in IndexedDB, etc.

Do you know what is the reason for this? Maybe privacy or security? If so, consider adding something at the end of the paragraph like "...but no browsers support this for privacy reasons".

Author: Yeah, there's an excellent piece on this here: https://web.dev/push-notifications-subscribing-a-user/#uservisibleonly-options .

From my reading, in Chrome anyway, if the SW has not shown a notification by the time `waitUntil()` resolves, then the browser shows some generic notification. I haven't tested this though! Maybe I should. It's hard to test though.

Anyway, added a bit here: https://github.com/mdn/content/pull/25305/commits/881aaa5e53cb4b838f1ff8a904af3feb6442ba31.
